File: 2024/puzzles/day1/day1.go
package day1

import (
	_ "embed"
	"slices"
	"strings"

	"github.com/KhachikAstoyan/advent-of-code/2024/utils"
)

type Day1 struct{}

//go:embed input.txt
var input string

func (d Day1) Name() string {
	return "Day 1: Historian Hysteria"
}

func (d Day1) Part1() uint64 {
	left, right := parseLists(strings.Split(input, "\n"))
	slices.Sort(left)
	slices.Sort(right)

	answer := uint64(0)
	for i := 0; i < len(left); i++ {
		answer += utils.AbsDiff(left[i], right[i])
	}

	return answer
}

func (d Day1) Part2() uint64 {
	left, right := parseLists(strings.Split(input, "\n"))
	m := make(map[uint64]uint64)

	for _, num := range left {
		m[num] = 0
	}

	for _, num := range right {
		if _, ok := m[num]; ok {
			m[num]++
		}
	}

	similarity := uint64(0)
	for num, count := range m {
		similarity += num * count
	}

	return similarity
}

func parseLists(lines []string) ([]uint64, []uint64) {
	n := len(lines)
	left := make([]uint64, n)
	right := make([]uint64, n)

	for i, line := range lines {
		parts := strings.Fields(line)

		leftNum := utils.ParseInt(parts[0], 10, 64)
		rightNum := utils.ParseInt(parts[1], 10, 64)

		left[i] = uint64(leftNum)
		right[i] = uint64(rightNum)
	}

	return left, right
}

File: 2024/src/days/day1.rs
use std::collections::HashMap;

use super::Day;

pub struct Day1 {}

impl Day for Day1 {
    fn name(&self) -> String {
        String::from("Day 1: Historian Hysteria")
    }

    // Correct answer: 2285373
    fn part_one(&self, input: &str) -> u64 {
        let (mut left, mut right) = parse_lists(input);
        left.sort();
        right.sort();

        let mut answer: u64 = 0;
        for i in 0..left.len() {
            answer += left[i].abs_diff(right[i]);
        }

        answer
    }

    // Correct answer: 21142653 
    fn part_two(&self, input: &str) -> u64 {
        let (left, right) = parse_lists(input);
        let mut map: HashMap<u64, u64> = HashMap::new();
        left.iter().for_each(|num| {
             map.entry(*num).or_insert(0);
        });

        right.iter().for_each(|num| {
            let current_entry = map.get(num);
            if current_entry.is_some() {
                map.insert(*num, current_entry.unwrap() + 1);
            }
        });

        let mut similarity_score = 0;
        for (num, count) in map.into_iter() {
            similarity_score += num * count;
        }
        
        similarity_score
    }
}

fn parse_lists(input: &str) -> (Vec<u64>, Vec<u64>) {
    let mut left: Vec<u64> = vec![];
    let mut right: Vec<u64> = vec![];

    input.lines().for_each(|line| {
        let mut split = line.split_whitespace();
        let left_num: u64 = split.next().unwrap().parse().unwrap();
        let right_num: u64 = split.next().unwrap().parse().unwrap();

        left.push(left_num);
        right.push(right_num);
    });

    return (left, right)
}
